daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/PackageInterfaceSpec.scala

```diff
@@ -0,0 +1,140 @@
+package com.daml.lf
+package language
+
+import com.daml.lf.data.Ref.Identifier
+import com.daml.lf.data._
+import com.daml.lf.language.Ast._
+import com.daml.lf.language.PackageInterface.{DataEnumInfo, DataRecordInfo, DataVariantInfo}
+import org.scalatest.{Assertion, Inside}
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.prop._
+import org.scalatest.wordspec.AnyWordSpec
+
+class PackageInterfaceSpec
+    extends AnyWordSpec
+    with Matchers
+    with TableDrivenPropertyChecks
+    with Inside {
+
+  import com.daml.lf.testing.parser.Implicits._
+  import com.daml.lf.transaction.test.TransactionBuilder.Implicits.{defaultPackageId => _, _}
+
+  private[this] implicit val defaultPackageId: Ref.PackageId =
+    defaultParserParameters.defaultPackageId
+
+  private[this] val pkg =
+    p"""
+        module Mod {
+
+          val unit: Unit = ();
+          
+          record @serializable Tuple X Y = { fst: X, snd: Y };
+          variant @serializable Either a b = Left : a | Right : b ;
+          enum @serializable Color = Red | Green | Blue;
+
+          record @serializable Contract = {};
+          template (this : Contract) =  {
+              precondition True,
+              signatories Nil @Party,
+              observers Nil @Party,
+              agreement "Agreement",
+              choices { }
+            };
+
+       }
+    """
+
+  private[this] val interface = PackageInterface(Map(defaultPackageId -> pkg))
+
+  private[this] def test[X, Y](
+      description: String,
+      lookup: X => Either[LookupError, Y],
+      toContext: X => Reference,
+  )(nonErrorCase: X, validateSuccess: PartialFunction[Y, Assertion], errorCases: (X, Reference)*) =
+    s"Lookup$description" should {
+
+      s"succeed on known ${description.toLowerCase()}" in {
+        inside(lookup(nonErrorCase)) { case Right(success) => inside(success)(validateSuccess) }
+      }
+
+      s"fail on unknown ${description.toLowerCase()}" in {
+        val table = Table("input" -> "reference", errorCases: _*)
+
+        forEvery(table) { (input, expectedNotFound) =>
+          assume(input != nonErrorCase)
+          inside(lookup(input)) { case Left(LookupError(notFound, context)) =>
+            notFound shouldBe expectedNotFound
+            context shouldBe toContext(input)
+          }
+        }
+      }
+    }
+
+  test("DataRecord", interface.lookupDataRecord, Reference.DataRecord)(
+    "Mod:Tuple",
+    { case DataRecordInfo(dataType, DataRecord(variants)) =>
```

Comment:
might be easier to read with named arguments.

Reply:
named- argument are not very nice with VarArgs function.